### PR TITLE
[8.0] fix: interacting with CEs using tokens

### PR DESCRIPTION
**Pull request by aldbr**
This PR aims at enhancing/fixing the interactions with CEs using tokens.

Improvement: 
- support for the `aud` claim in the `SiteDirector`: an access token targets a specific CE (limit its impact if stolen).
  - add the `audienceName` attribute in `ComputingElement`.

Fix:
- `TokenManager(Client/Handler/Utilities)`: 
  - change the `scope` type (`str` -> `list[str]`)
  - pass the `IdProvider` object instead of its name to `getCachedKey()`
  - initialize `tokensCache` and `idps` before the database. Note: this is likely a ugly hack, but if we only use the `client_credentials` flow, we do not need the database: by moving `tokensCache` and `idps` before the initialization of the database, the service can work without it (it gets an error at the initialization but it works).
- `Grid`:
  - move the proxy initialization in `HTCondorCE`. By the way `Grid` is only used by `HTCondorCE` now.
- `WMSUtilities`:
  - make `setPilotCredentials()` compatible with the `client_credentials` flow.
- `IdProviderFactory`:
  - move a block of code related to the DIRAC AS in a condition branch so that it is not needed if we only work with external IdPs to interact with CEs.

Plus, I added a few tests (`OAuth2IdProvider.fetchToken()`, `TokenManagerUtilities`).   

BEGINRELEASENOTES
*Resources
NEW: audienceName attribute in CEs
*WorkloadManagementSystem
NEW: audience support in SiteDirector
*FrameworkSyste
FIX: fix TokenManager(Client/Handler/Utilities)
ENDRELEASENOTES

### CODE REVIEW

**Comment by fstagni** [comment]
You committed an empty src/DIRAC/WorkloadManagementSystem/private/test/Test_ConfigHelper.py, was that on purpose?

**Comment by aldbr** [comment]
> You committed an empty src/DIRAC/WorkloadManagementSystem/private/test/Test_ConfigHelper.py, was that on purpose?

No, I've just removed it.

File: src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py [hunk 1]
```diff
@@ -465,12 +465,20 @@ def submitPilots(self):
...
         """
-        result = gTokenManager.getToken(userGroup=self.pilotGroup, requiredTimeLeft=600)
+        if not audience:
+            return S_ERROR("Audience is not defined")
+
+        if not scope:
+            scope = ["compute.cancel", "compute.create", "compute.read", "compute.cancel"]
+
+        result = gTokenManager.getToken(userGroup=self.pilotGroup, requiredTimeLeft=600, scope=scope, audience=audience)
         return result
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py]:
```suggestion
        return gTokenManager.getToken(userGroup=self.pilotGroup, requiredTimeLeft=600, scope=scope, audience=audience)
```

File: src/DIRAC/WorkloadManagementSystem/Service/WMSUtilities.py [hunk 2]
```diff
@@ -96,10 +75,15 @@ def setPilotCredentials(ce, pilotDict):
     :return: S_OK/S_ERROR
     """
     if "Token" in ce.ceParameters.get("Tag", []):
-        result = getPilotToken(pilotDict)
+        result = gTokenManager.getToken(
+            userGroup=pilotDict["OwnerGroup"],
+            scope=["compute.cancel", "compute.create", "compute.modify", "compute.read"],
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Service/WMSUtilities.py]: Only hardcoded?

Comment (aldbr) [reply src/DIRAC/WorkloadManagementSystem/Service/WMSUtilities.py]: Do you have any suggestion?

Comment (fstagni) [reply src/DIRAC/WorkloadManagementSystem/Service/WMSUtilities.py]: At a minimum I would put a "constant" `SCOPES=["compute.cancel", "compute.create", "compute.modify", "compute.read"]` in order to be less hidden. And maybe a CS entry...

Comment (aldbr) [reply src/DIRAC/WorkloadManagementSystem/Service/WMSUtilities.py]: What about this? https://github.com/DIRACGrid/DIRAC/pull/7062/commits/6c2b4f8f8423c02723f32ad9e53e0e293d5504fb

**Review by fstagni — APPROVED** [review]
OK for me but @atsareg should review this one too.

**Pull request merged**